src/librustc_trans/trans/intrinsic.rs

```diff
@@ -1270,3 +1323,335 @@ fn get_rust_try_fn<'a, 'tcx>(fcx: &FunctionContext<'a, 'tcx>,
+        let in_len = arg_tys[0].simd_size(tcx);
+        let out_len = ret_ty.simd_size(tcx);
+        require!(out_len == n,
+                 "SIMD shuffle intrinsic monomorphized with return type of length {} (expected {})",
+                 out_len, n);
+        require!(arg_tys[0].simd_type(tcx) == ret_ty.simd_type(tcx),
+                 "SIMD shuffle intrinsic monomorphized with different \
+                  input and return element types");
+
+        let total_len = in_len as u64 * 2;
+
+        let vector = match args {
+            Some(args) => &args[2],
+            None => bcx.sess().span_bug(call_info.span,
+                                        "intrinsic call with unexpected argument shape"),
+        };
+        let vector = consts::const_expr(bcx.ccx(), vector, tcx.mk_substs(substs), None).0;
+
+        let indices: Option<Vec<_>> = (0..n)
+            .map(|i| {
+                let arg_idx = i;
+                let val = const_get_elt(bcx.ccx(), vector, &[i as libc::c_uint]);
+                let c = const_to_opt_uint(val);
+                match c {
+                    None => {
+                        bcx.sess().span_err(call_info.span,
+                                            &format!("SIMD shuffle intrinsic argument #{} \
+                                                      is not a constant",
+                                                     arg_idx));
+                        None
+                    }
+                    Some(idx) if idx >= total_len => {
+                        bcx.sess().span_err(call_info.span,
+                                            &format!("SIMD shuffle intrinsic argument #{} \
+                                                      is out of bounds (limit {})",
+                                                     arg_idx, total_len));
+                        None
+                    }
+                    Some(idx) => Some(C_i32(bcx.ccx(), idx as i32)),
+                }
+            })
+            .collect();
+        let indices = match indices {
+            Some(i) => i,
+            None => return C_null(llret_ty)
+        };
+
+        return ShuffleVector(bcx, llargs[0], llargs[1], C_vector(&indices))
+    }
+
+    if name == "simd_insert" {
+        require!(arg_tys[0].is_simd(),
+                 "SIMD insert intrinsic monomorphized for non-SIMD input type");
+
+        let elem_ty = arg_tys[0].simd_type(tcx);
+        require!(arg_tys[2] == elem_ty,
+                 "SIMD insert intrinsic monomorphized with inserted type not SIMD element type");
+        return InsertElement(bcx, llargs[0], llargs[2], llargs[1])
+    }
+    if name == "simd_extract" {
+        require!(arg_tys[0].is_simd(),
+                 "SIMD insert intrinsic monomorphized for non-SIMD input type");
+
+        let elem_ty = arg_tys[0].simd_type(tcx);
+        require!(ret_ty == elem_ty,
+                 "SIMD insert intrinsic monomorphized with returned type not SIMD element type");
```

Comment:
s/insert/extract/ 

Although I'd also be in favor of just unifying these two branches to reduce boilerplate